src/dataflow/src/server.rs

```diff
@@ -218,19 +217,216 @@ pub fn serve(config: Config) -> Result<WorkerGuards<()>, String> {
+        entry.push((timestamp, offset));
+    }
+
+    /// Get the timestamp assignment for `(partition, offset)` if it is known.
+    ///
+    /// This function returns the timestamp and the maximum offset for which it is
+    /// valid.
+    fn get_binding(
+        &self,
+        partition: &PartitionId,
+        offset: MzOffset,
+    ) -> Option<(Timestamp, MzOffset)> {
+        if !self.partitions.contains_key(partition) {
+            return None;
+        }
+
+        let entries = self.partitions.get(partition).expect("known to exist");
+
+        for (ts, max_offset) in entries {
+            if offset <= *max_offset {
+                return Some((*ts, *max_offset));
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns the maximum timestamp that is bound on all
+    /// of the partitions (e.g. the smallest timestamp that every
+    /// single partition knows about). We can guarantee that all
+    /// subsequent updates will be in advance of this timestamp.
+    fn get_upper(&self) -> Antichain<Timestamp> {
+        let mut upper = Antichain::new();
+
+        for (_, entries) in self.partitions.iter() {
+            if let Some((timestamp, _)) = entries.last() {
+                upper.insert(*timestamp);
+            }
+        }
+
+        upper
+    }
+
+    /// Returns the list of partitions this source knows about.
+    ///
+    /// TODO(rkhaitan): this function feels like a hack, both in the API of having
+    /// the source instances ask for the list of known partitions and in allocating
+    /// a vector to answer that question.
+    fn partitions(&self) -> Vec<PartitionId> {
+        self.partitions
+            .iter()
+            .map(|(pid, _)| pid)
+            .cloned()
+            .collect()
+    }
+}
+
+/// A wrapper that allows multiple source instances to share a `TimestampBindingBox`
+/// and hold back its compaction.
+pub struct TimestampBindingRc {
+    wrapper: Rc<RefCell<TimestampBindingBox>>,
+    compaction_frontier: Antichain<Timestamp>,
+}
+
+impl TimestampBindingRc {
+    pub fn new() -> Self {
+        let wrapper = Rc::new(RefCell::new(TimestampBindingBox::new()));
+
+        let ret = Self {
+            wrapper: wrapper.clone(),
+            compaction_frontier: wrapper.borrow().compaction_frontier.frontier().to_owned(),
+        };
+
+        ret
+    }
+
+    pub fn set_compaction_frontier(&mut self, new_frontier: AntichainRef<Timestamp>) {
+        self.wrapper
+            .borrow_mut()
+            .adjust_compaction_frontier(self.compaction_frontier.borrow(), new_frontier);
+        self.compaction_frontier = new_frontier.to_owned();
+    }
+
+    fn compact(&self) {
+        self.wrapper.borrow_mut().compact();
+    }
+
+    fn add_binding(&self, partition: PartitionId, timestamp: Timestamp, offset: MzOffset) {
+        self.wrapper
+            .borrow_mut()
+            .add_binding(partition, timestamp, offset);
+    }
+
+    pub fn get_binding(
+        &self,
```

Comment:
There are a bunch of docs on the `Box` type, but also the warning not to use it. It seems like the docs are probably better here if anywhere?

Reply:
docs moved + everything moved to a new file